### PR TITLE
Create availability controller

### LINKED ISSUE #45: BE: Availability controller
Changes include the following:

- Model for availability needs to be completed first.
- New routes for availability.
   - `/availability` <- to create a schedule
   - `/availability/:scheduleId`
   - `/availability/active` <- This route should just get their active schedule
   - `/availability` <- All availability schedule for the current signed in user.
   - `/availability/:scheduleId/activate` <- This will set the active schedule id on the profile model for that pet sitter.
- Controllers for all these routes. Think about middleware too and how it can help us, particularly with fetching profile, **input validation**.
- Make sure to keep things DRY (Don't Repeat Yourself) in this controller.

### PR DESCRIPTION
**Pull request by panwaramita**
### What this PR does (required):
- Create new routes for availability.
/availability <- to create a schedule
/availability/:scheduleId
/availability/active <- This route should just get their active schedule
/availability <- All availability schedule for the current signed-in user.
/availability/:scheduleId/activate <- This will set the active schedule id on the profile model for that pet sitter.

-Create controller for each route

Issue Link : Closes #45 

### Any information needed to test this feature (required):
Since the UI is not available the routes can be tested using the below paths in postman(https://localhost:3001):
1. Create the schedule /availability.
2. Get schedule based on ID /availability/:scheduleId.
3. Get an active schedule /availability/active.
4. Get all the schedule for user /availability /all
5. Set active schedule for user /availability/:scheduleId/activate

### CODE REVIEW

File: server/controllers/availability.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            res.send(schedule);
+        } else {
+            res.status(204).json({});
+            throw new Error("No Schedule found");
+        }
+      });
+
+// @route GET /availability/active
+// @desc Get active schedule
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/availability.js:71]: :+1:

**Review by ezeilo-su — CHANGES_REQUESTED** [review]
Hey @panwaramita, good progress so far. You need to apply proper code formatting to your code. Use the recommended linters setup to auto-format all your code.
Happy coding!

File: server/controllers/availability.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+   const scheduleId=req.params.scheduleId;  
+   const schedule = await Availability.findOne({ _id: scheduleId })
+        if (schedule) {
+            res.status(200);
+            res.send(schedule);
+        } else {
+            res.status(204).json({});
+            throw new Error("No Schedule found");
+        }
+        });
```
[the remark this answers is not in the document]
Comment (panwaramita) [reply server/controllers/availability.js:52]: Thanks for the comments. Done the formatting.

File: server/controllers/availability.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    const newSchedule = new Availability.create({
+        profileId,
+        name,
+        days,
+      });    
+      if (newSchedule) {
+          if(!req.profile.activeSchedule)
+       {
+        req.profile.set({ activeSchedule: newSchedule.id });
+        await req.profile.save();
```
[the remark this answers is not in the document]
Comment (panwaramita) [reply server/controllers/availability.js:26]: Thanks for the comments. Done the formatting.

File: server/controllers/availability.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    if (!req.profile.activeSchedule) {
+      req.profile.set({ activeSchedule: newSchedule.id });
+      await req.profile.save();
+    }
+    res.status(201).json({
+      success: {
+        newSchedule
+      }
+    });
+  } else {
```
Comment (rajivtitus) [inline server/controllers/availability.js:31]: I don't think we'll reach this else block if the above async operation fails, the `asyncHandler` will instead forward the error to our error middleware

Ditto where applicable

File: server/controllers/availability.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+// @desc Get schedule
+// @access Private
+exports.getScheduleId = asyncHandler(async (req, res, next) => {
+  const scheduleId = req.params.scheduleId;
+  const schedule = await Availability.findOne({ _id: scheduleId })
+  if (schedule) {
+    res.status(200);
+    res.send(schedule);
+  } else {
+    res.status(204).json({});
```
Comment (rajivtitus) [inline server/controllers/availability.js:47]:
```suggestion
    res.status(404).json({});
```

File: server/controllers/availability.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    throw new Error("Invalid data");
+  }
+});
+
+// @route GET /availability/:scheduleId
+// @desc Get schedule
+// @access Private
+exports.getScheduleId = asyncHandler(async (req, res, next) => {
+  const scheduleId = req.params.scheduleId;
+  const schedule = await Availability.findOne({ _id: scheduleId })
```
Comment (rajivtitus) [inline server/controllers/availability.js:42]: Let's also make sure that only the logged in user can access their availability schedule

File: server/controllers/availability.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// @route GET /availability
+// @desc Get schedule
+// @access Private
+exports.getSchedule = asyncHandler(async (req, res, next) => {
+  const profileId = req.profile.id;
+  const schedule = await Availability.find({ petSitterId: profileId })
+  if (schedule) {
+    res.status(200);
+    res.send(schedule);
+  } else {
```
Comment (rajivtitus) [inline server/controllers/availability.js:61]: We don't need this else bock as `schedule` is a list of availability docs or an empty list (both truthy values)

File: server/controllers/availability.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+// @desc Set active schedule
+// @access Private
+exports.setActiveSchedule = asyncHandler(async (req, res, next) => {
+  const scheduleId = req.params.scheduleId;
+  const activeSchedule = await Profile.findOne({ activeSchedule: scheduleId });
+  if (!activeSchedule) {
+    res.status(404);
+    throw new Error("Profile doesn't exist");
+  }
+  activeSchedule.set(req.body);
```
Comment (rajivtitus) [inline server/controllers/availability.js:92]: Only the logged in user should be able to set their active schedule

File: server/validate.js [hunk 9]
```diff
@@ -30,3 +30,17 @@ exports.validateLogin = [
     next();
   }
 ];
+
+exports.validateDays=[
```
Comment (rajivtitus) [inline server/validate.js:34]: 👍🏽

File: server/models/availability.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
+const mongoose = require("mongoose");
```
Comment (rajivtitus) [inline server/models/availability.js:1]: Please adjust the base of the PR/rebase to only reflect changes pertaining to this PR

File: server/controllers/availability.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
+const Availability = require("../models/availability");
+const asyncHandler = require("express-async-handler");
+const Profile = require("../models/Profile");
+
+// @route POST /schedule/
+// @desc Create a schedule
+// @access Public
+exports.createSchedule = asyncHandler(async (req, res, next) => {
+  const profileId = req.profile.id;
```
Comment (rajivtitus) [inline server/controllers/availability.js:9]: Question: Does the `req` object contain the profile id? As far as my knowledge goes, we only have access to the user id and not the profile?

**Review by rajivtitus — COMMENTED** [review]
Good work Amita! Some comments about security in the controllers, please have a look and let me know if you have any questions :)

File: server/controllers/availability.js (hunk 11 above)
Comment (panwaramita) [reply server/controllers/availability.js:9]: No,req obeject does not contain profile id. Made the changes to get the user id from req object and from user id get the profile id from profile model.

File: server/controllers/availability.js (hunk 8 above)
Comment (panwaramita) [reply server/controllers/availability.js:92]: Done the changes

File: server/controllers/availability.js (hunk 7 above)
Comment (panwaramita) [reply server/controllers/availability.js:61]: removed the else block

File: server/controllers/availability.js (hunk 6 above)
Comment (panwaramita) [reply server/controllers/availability.js:42]: done the changes.

File: server/controllers/availability.js (hunk 5 above)
Comment (panwaramita) [reply server/controllers/availability.js:47]: Done the changes

File: server/controllers/availability.js (hunk 4 above)
Comment (panwaramita) [reply server/controllers/availability.js:31]: Done the changes

File: server/controllers/availability.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  const { name, days } = req.body;
+  const schedule = await Availability.findOne({ name, petSitterId: req.profile.id });
+  if (schedule) {
+    res.status(400);
+    throw new Error("Schedule name already exist");
+  }
+  const newSchedule = new Availability.create({
```
[the remark this answers is not in the document]
Comment (panwaramita) [reply server/controllers/availability.js:16]: Done the changes

File: server/controllers/availability.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+// @desc Create a schedule
+// @access Public
+exports.createSchedule = asyncHandler(async (req, res, next) => {
+  const profile = await User.findById(req.user.id, "profile");
+
+  if (!profile) {
+    res.status(401);
+    throw new Error("Not authorized");
+  }
+  const profileId = await Profile.findOne({ userId: req.user.id });
```
Comment (rajivtitus) [inline server/controllers/availability.js:15]:
```suggestion
  const profile = await Profile.findOne({ userId: req.user.id });
```

File: server/controllers/availability.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  const id = profileId._id;
```
Comment (rajivtitus) [inline server/controllers/availability.js:16]:
```suggestion
  const profileId = profile._id;
```
You can change this throughout

File: server/controllers/availability.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    res.status(400);
+    throw new Error("Schedule name already exist");
+  }
+  const newSchedule = Availability.create({
+    petSitterId: id,
+    name,
+    days,
+  });
+  if (newSchedule) {
+    if (!profileId.activeSchedule) {
```
Comment (rajivtitus) [inline server/controllers/availability.js:29]: 💯

File: server/controllers/availability.js [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+});
+
+// @route GET /availability
+// @desc Get schedule
+// @access Private
+exports.getSchedule = asyncHandler(async (req, res, next) => {
+  const profileId = await Profile.findOne({ userId: req.user.id });
+  const id = profileId._id;
+  const schedule = await Availability.find({ petSitterId: id })
+  if (schedule) {
```
Comment (rajivtitus) [inline server/controllers/availability.js:63]: We don't need this `if` clause as the mongoose query `Model.find()` returns an empty list if no matching docs are found (which is a truthy value)

File: server/controllers/availability.js (hunk 8 above)
Comment (rajivtitus) [reply server/controllers/availability.js:97]: @panwaramita Please address this comment as well

**Review by rajivtitus — COMMENTED** [review]
Good work with the changes. A couple of comments yet to be addressed 😄

File: server/controllers/availability.js (hunk 16 above)
Comment (panwaramita) [reply server/controllers/availability.js:63]: Removed the if clause.

File: server/controllers/availability.js [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    res.status(404).json({});
+    throw new Error("No Schedule found");
+  }
+});
+
+// @route GET /availability
+// @desc Get schedule
+// @access Private
+exports.getSchedule = asyncHandler(async (req, res, next) => {
+  const profileId = await Profile.findOne({ userId: req.user.id });
```
[the remark this answers is not in the document]
Comment (panwaramita) [reply server/controllers/availability.js:60]: changed the profileId to profile.

File: server/controllers/availability.js (hunk 14 above)
Comment (panwaramita) [reply server/controllers/availability.js:16]: changed from id to ProfileId and profileId to profile throughout the controller

File: server/controllers/availability.js (hunk 13 above)
Comment (panwaramita) [reply server/controllers/availability.js:15]: Done the changes

File: server/controllers/availability.js (hunk 6 above)
Comment (panwaramita) [reply server/controllers/availability.js:46]: Added a clause to check if the req.user.id is present in the profile. If not throw an error "not authorized". else allow the user to access their schedule.

File: server/controllers/availability.js (hunk 8 above)
Comment (panwaramita) [reply server/controllers/availability.js:97]: Added a clause to check if the req.user.id is present in the profile. If not throw an error "not authorized". else allow the user to access their schedule.

File: server/controllers/availability.js [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+  if ( !profile )
+  {
+    res.status( 401 );
+    throw new Error( "Not authorized" );
+  }
+  const active = await Availability.findOne( { _id: profile.activeSchedule } );
+  if ( active )
+  {
+    res.status( 200 );
+    res.send( schedule );
```
Comment (rajivtitus) [inline server/controllers/availability.js:96]: `schedule` does not exist within this controller

File: server/controllers/availability.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+// @access Private
+exports.setActiveSchedule = asyncHandler( async ( req, res, next ) => {
+  const profile = await Profile.findOne( { userId: req.user.id } );
+  if ( !profile )
+  {
+    res.status( 401 );
+    throw new Error( "Not authorized" );
+  }
+  const scheduleId = req.params.scheduleId;
+  const activeSchedule = await Profile.findOne( { activeSchedule: scheduleId } );
```
Comment (rajivtitus) [inline server/controllers/availability.js:111]:
```suggestion
  const profile = await Profile.findOne( { activeSchedule: scheduleId } );
```
This query will return the entire profile document

Comment (rajivtitus) [inline server/controllers/availability.js:111]: Aren't we supposed to query the Availability model here?

File: server/controllers/availability.js [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+// @access Private
+exports.getScheduleId = asyncHandler( async ( req, res, next ) => {
+  const profile = await Profile.findOne( { userId: req.user.id } );
+  if ( !profile )
+  {
+    res.status( 401 );
+    throw new Error( "Not authorized" );
+  }
+  const scheduleId = req.params.scheduleId;
+  const schedule = await Availability.findOne( { _id: scheduleId } )
```
[the remark this answers is not in the document]
Comment (panwaramita) [reply server/controllers/availability.js:54]: included a condition to check if the schedule exists and petsitterid equals to scheduleId.

File: server/controllers/availability.js (hunk 18 above)
Comment (panwaramita) [reply server/controllers/availability.js:96]: Done the changes

File: server/controllers/availability.js (hunk 19 above)
Comment (panwaramita) [reply server/controllers/availability.js:111]: Done changes.The activeSchedule field is present in the profile document. Therefore, when the user makes a specific schedule active the profile sctiveSchedule value will be updated.

**Pull request merged**